pytest_operator/plugin.py

```diff
@@ -393,6 +494,98 @@ async def build_charms(self, *charm_paths):
 
+    @staticmethod
+    def _charm_file_resources(built_charm: Path):
+        if not built_charm.exists():
+            raise FileNotFoundError(f"Failed to locate built charm {built_charm}")
+
+        charm_path = ZipPath(built_charm)
+        metadata_path = charm_path / "metadata.yaml"
+        resources = yaml.safe_load(metadata_path.read_text())["resources"]
+
+        return {
+            name: Path(resource.get("filename"))
+            for name, resource in resources.items()
+            if resource.get("type") == "file"
+        }
+
+    def arch_specific_resources(self, built_charm: Path):
+        """
+        Discover architecture specific charm file resources by presence of
+        arch names in the resource.  If there is a resources that shares the same
+        base and tail of another arch specific resource, assume that it's an
+        amd64 specific resource.
+        """
+        resources = self._charm_file_resources(built_charm)
+        arch_based = re.compile(r"^(\S+)(?:-(amd64|arm64|s390x))(\S+)?")
+        arch_resources = (arch_based.match(rsc) for rsc in resources)
+        arch_resources = {
+            rsc.string: Arch(*rsc.groups()) for rsc in arch_resources if rsc
+        }
+
+        # some resources are arch specific but don't include amd64 in its name
+        # they'll be identified as being named <base><tail> where there is
+        # another resource with <base>-<arch><tail>
+        potentials = {(rsc.base, rsc.tail) for rsc in arch_resources.values()}
+        for base, tail in potentials:
+            amd64_rsc = f"{base}{tail or ''}"
+            if resources.get(amd64_rsc):
+                arch_resources[amd64_rsc] = Arch(base, "amd64", tail)
+        return arch_resources
+
+    async def build_resources(self, build_script: Path):
+        build_script = build_script.absolute()
+        if not build_script.exists():
+            raise FileNotFoundError(
+                f"Failed to locate resource build script {build_script}"
+            )
+
+        log.info("Build Resources...")
+        dst_dir = self.tmp_path / "resources"
+        dst_dir.mkdir(exist_ok=True)
+        rc, stdout, stderr = await self.run(
+            *shlex.split(f"sudo {build_script}"), cwd=dst_dir, check=False
+        )
+        if rc != 0:
+            log.warning(f"{build_script} failed: {(stderr or stdout).strip()}")
+        return list(dst_dir.glob("*.*"))
+
+    @staticmethod
+    def _charm_name(built_charm: Path):
+        if not built_charm.exists():
+            raise FileNotFoundError(f"Failed to locate built charm {built_charm}")
+
+        charm_path = ZipPath(built_charm)
+        metadata_path = charm_path / "metadata.yaml"
+        return yaml.safe_load(metadata_path.read_text())["name"]
+
+    async def download_resources(
+        self,
+        built_charm: Path,
+        owner="",
+        channel="edge",
+        filter_in=lambda rsc: True,
+        name=lambda rsc, path: path,
```

Comment:
I don't find `name` very descriptive.  What about this instead:

```suggestion
        map_filename=lambda rsc, filename: filename,
```